### PR TITLE
feat: add the ability to configure artifact download link expiry

### PR DESCRIPTION
**Pull request by HumairAK**
## The issue resolved by this Pull Request:
Part of https://issues.redhat.com/browse/RHOAIENG-8225

Depends on:  https://github.com/opendatahub-io/data-science-pipelines/pull/62

## Description of your changes:

This allows a user to configure the expiry time (in seconds) of an artifact download link via the DSPA. Default is 15 seconds.

## Testing instructions

1. Deploy DSPO: Use this one https://github.com/opendatahub-io/data-science-pipelines-operator/pull/670
2. Deploy DSPA with `.spec.apiServer.artifactSignedURLExpirySeconds` set to some value integer.
3. Confirm this value is set in the dsp server deployment as: `SIGNED_URL_EXPIRY_TIME_SECONDS`

## Checklist
- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

**Comment by dsp-developers** [comment]
A new image has been built to help with testing out this PR: `quay.io/opendatahub/data-science-pipelines-operator:pr-670`
An OCP cluster where you are logged in as cluster admin is required.

To use this image run the following:

```bash
cd $(mktemp -d)
git clone git@github.com:opendatahub-io/data-science-pipelines-operator.git
cd data-science-pipelines-operator/
git fetch origin pull/670/head
git checkout -b pullrequest 8a74a7b680e73bd431f9f18dfe12ae540ab26524
oc new-project opendatahub
make deploy IMG="quay.io/opendatahub/data-science-pipelines-operator:pr-670"
```

More instructions [here](https://github.com/opendatahub-io/data-science-pipelines-operator#deploy-dsp-instance) on how to deploy and test a Data Science Pipelines Application.

**Comment by dsp-developers** [comment]
Change to PR detected. A new PR build was completed.
  A new image has been built to help with testing out this PR: `quay.io/opendatahub/data-science-pipelines-operator:pr-670`

**Review by hbelmiro — COMMENTED** [review]
/lgtm

File: config/samples/v2/dspa-simple/dspa_simple.yaml [hunk 1]
```diff
@@ -7,6 +7,9 @@ spec:
   apiServer:
     enableSamplePipeline: true
   objectStorage:
+    # Need to enable this for artifact download links to work
```
Comment (gmfrasca) [inline config/samples/v2/dspa-simple/dspa_simple.yaml:10]: if this is the case, should we make the objStore route enabled by default?  (outside scope of this ticket, but if so we should open a ticket to track/follow up on)

**Comment by gmfrasca** [comment]
/approve 

but also 

/hold

since I believe this needs to go in around the same time as https://github.com/opendatahub-io/data-science-pipelines/pull/62

also giving @HumairAK a chance to see above [non-blocking comment](https://github.com/opendatahub-io/data-science-pipelines-operator/pull/670#discussion_r1674977393), so allowing for a self-unhold

File: config/samples/v2/dspa-simple/dspa_simple.yaml (hunk 1 above)
Comment (HumairAK) [reply config/samples/v2/dspa-simple/dspa_simple.yaml:10]: I don't have a strong opinion on this, this field itself is misplaced (it should be under minio), it only applies to minio which is a dev feature, so I don't think it matters that much.

**Comment by gregsheremeta** [comment]
/hold

need to firm up the API in https://github.com/opendatahub-io/data-science-pipelines/pull/62

(I'm not sure all the references to `share_url` should remain)

**Comment by gregsheremeta** [comment]
can you change the commit message and the PR title?

> feat: configure artifact download link expiry

(the artifacts aren't expired! just their links ;) )

File: api/v1alpha1/dspipeline_types.go [hunk 2]
```diff
@@ -141,6 +141,13 @@ type APIServer struct {
 	// pipeline server and user executor pods
 	// +kubebuilder:validation:Optional
 	CABundleFileName string `json:"caBundleFileName"`
+
+	// The expiry time (seconds) for artifact download links when
+	// querying the dsp server via /apis/v2beta1/artifacts/{id}?share_url=true
+	// Default: 15
+	// +kubebuilder:default:=15
+	// +kubebuilder:validation:Optional
+	ArtifactSignedURLExpirySeconds *int `json:"artifactSignedURLExpirySeconds"`
```
Comment (gregsheremeta) [inline api/v1alpha1/dspipeline_types.go:150]: this is refered to as "presigned" URLs everywhere I've looked


```suggestion
	ArtifactPresignedURLExpirySeconds *int `json:"artifactPresignedURLExpirySeconds"`
```

File: config/internal/apiserver/default/deployment.yaml.tmpl [hunk 3]
```diff
@@ -116,6 +116,8 @@ spec:
               value: "ds-pipeline-{{.Name}}.{{.Namespace}}.svc.cluster.local"
             - name: ML_PIPELINE_SERVICE_PORT_GRPC
               value: "8887"
+            - name: SIGNED_URL_EXPIRY_TIME_SECONDS
```
Comment (gregsheremeta) [inline config/internal/apiserver/default/deployment.yaml.tmpl:119]:
```suggestion
            - name: PRESIGNED_URL_EXPIRY_TIME_SECONDS
```

File: controllers/config/defaults.go [hunk 4]
```diff
@@ -57,6 +57,8 @@ const (
 	DefaultDBSecretKey        = "password"
 	GeneratedDBPasswordLength = 12
 
+	DefaultSignedUrlExpiryTimeSeconds = 15
```
Comment (gregsheremeta) [inline controllers/config/defaults.go:60]:
```suggestion
	DefaultPresignedUrlExpiryTimeSeconds = 15
```

File: controllers/testdata/declarative/case_6/deploy/03_cr.yaml [hunk 5]
```diff
@@ -5,6 +5,7 @@
 # MLMD grpc server mounts the dspa cert and passes it into grpc server
 # When a user provides a caBundleFileMountPath, it will be used to mount the ca bundle
 # When a user provides ca bundle configmapkey, it will be used instead of default one
+# When a user provides a ArtifactSignedURLExpirySeconds, it will be used instead of default
```
Comment (gregsheremeta) [inline controllers/testdata/declarative/case_6/deploy/03_cr.yaml:8]:
```suggestion
# When a user provides a ArtifactPresignedURLExpirySeconds, it will be used instead of default
```

File: config/internal/apiserver/default/role_ds-pipeline.yaml.tmpl [hunk 6]
```diff
@@ -16,6 +16,13 @@ rules:
...
       - list
       - delete
+  - apiGroups:
+      - ""
+    resources:
+      - secrets
+    verbs:
+      - get
+      - list
   - apiGroups:
```
Comment (gregsheremeta) [inline config/internal/apiserver/default/role_ds-pipeline.yaml.tmpl:26]: why is this needed? I don't see any changes to code that gets or lists secrets.

File: config/internal/apiserver/default/deployment.yaml.tmpl (hunk 3 above)
Comment (HumairAK) [reply config/internal/apiserver/default/deployment.yaml.tmpl:119]: I used signed specifically because some object stores like GCS refer to it as "signed url" [[1]](https://cloud.google.com/storage/docs/access-control/signed-urls), s

File: config/internal/apiserver/default/role_ds-pipeline.yaml.tmpl (hunk 6 above)
Comment (HumairAK) [reply config/internal/apiserver/default/role_ds-pipeline.yaml.tmpl:26]: we have to fetch the object store creds for creating the s3 client to create the presigned url

File: controllers/config/defaults.go (hunk 4 above)
Comment (HumairAK) [reply controllers/config/defaults.go:60]: see https://github.com/opendatahub-io/data-science-pipelines-operator/pull/670#discussion_r1677187381

File: controllers/testdata/declarative/case_6/deploy/03_cr.yaml (hunk 5 above)
Comment (HumairAK) [reply controllers/testdata/declarative/case_6/deploy/03_cr.yaml:8]: see https://github.com/opendatahub-io/data-science-pipelines-operator/pull/670#discussion_r1677187381

File: api/v1alpha1/dspipeline_types.go (hunk 2 above)
Comment (HumairAK) [reply api/v1alpha1/dspipeline_types.go:150]: see https://github.com/opendatahub-io/data-science-pipelines-operator/pull/670#discussion_r1677187381

**Comment by dsp-developers** [comment]
Change to PR detected. A new PR build was completed.
  A new image has been built to help with testing out this PR: `quay.io/opendatahub/data-science-pipelines-operator:pr-670`
